### ISSUE TITLE
Add compartmental model example to Starsim

### ISSUE THREAD

**Issue by cliffckerr** [starsimhub/starsim#644]
Curious to see how far we can get using non-boolean state values, i.e. approximate a compartmental model.

**Comment by cliffckerr** [comment]
Pretty far :exploding_head: Can add as an advanced tutorial.
```py
"""
Example compartmental SIS model for Starsim
"""

import starsim as ss
import matplotlib.pyplot as plt


class compartmental_SIS(ss.Module):
    def __init__(self, pars=None, *args, **kwargs):
        super().__init__()
        self.define_pars(
            beta = ss.rate(0.8),
            init_prev = 0.01, # Not a distribution
            recovery = ss.rate(0.1), # Not a distribution
            waning = ss.rate(0.05),
            imm_boost = 1.0,
            use_immunity = True,
        )
        self.update_pars(pars=pars, *args, **kwargs)
        
        # Don't need to define states just use scalars
        self.N = 0
        self.S = 0
        self.I = 0
        self.immunity = 0
        return
    
    def init_post(self):
        """ Finish initialization """
        super().init_post()
        self.N = len(self.sim.people) # Could change to support a dynamic population size
        i0 = self.pars.init_prev
        self.S = self.N*(1-i0)
        self.I = self.N*i0
        self.immunity = i0
        return
        
    @property
    def rel_sus(self):
        return 1 - self.immunity
    
    def step(self):
        """ Carry out disease transmission logic """
        self.immunity *= (1 - self.pars.waning) # Update immunity from waning
        rel_sus = self.rel_sus if self.pars.use_immunity else 1.0
        infected = (self.S*self.I/self.N)*self.pars.beta*rel_sus
        recovered = self.I*self.pars.recovery
        net = infected - recovered
        self.S -= net
        self.I += net
        self.immunity += infected/self.N*self.pars.imm_boost # Update immunity from new infections
        return
    
    def init_results(self):
        """ Initialize results """
        super().init_results()
        self.results += ss.Result(self.name, 'S', self.npts, dtype=float, label='Susceptible')
        self.results += ss.Result(self.name, 'I', self.npts, dtype=float, label='Infectious')
        return

    def update_results(self):
        """ Store the current state """
        super().update_results()
        self.results['S'][self.ti] = self.S
        self.results['I'][self.ti] = self.I
        return 

    def plot(self):
        """ Default plot for SIS model """
        fig = plt.figure()
        res = self.results
        kw = dict(lw=3, alpha=0.7)
        for rkey in ['S', 'I']:
            plt.plot(self.timevec, res[rkey], label=res[rkey].label, **kw)
        plt.legend()
        plt.xlim(left=self.timevec[0])
        plt.ylim(bottom=0)
        plt.show()
        return fig
    

if __name__ == '__main__':
    sim = ss.Sim(dur=100, dt=0.1, diseases=compartmental_SIS())
    sim.run()
    sim.diseases.compartmental_sis.plot() 
```
![image](https://github.com/user-attachments/assets/2a46d6e6-88a8-4b89-8862-5f52c642f872)

**Comment by cliffckerr** [comment]
Closed by #654
